src/modm/driver/encoder/as5600.hpp

```diff
@@ -0,0 +1,332 @@
+	reset()
+	{
+		RF_BEGIN();
+
+		bool success = true;
+
+		/// The config registers span from 0x00 to 0x08
+		for (uint8_t reg = 0x00; reg < 0x08; reg += 2)
+		{
+			buffer[0] = reg;
+			buffer[1] = 0;
+			buffer[2] = 0;
+
+			this->transaction.configureWrite(buffer, 3);
+
+			RF_CALL(this->runTransaction());
+			success &= this->wasTransactionSuccessful();
+		}
+
+		RF_END_RETURN(success);
+	}
+
+	modm::ResumableResult<bool>
+	configure(Config_t config)
+	{
+		return write(Register::CONF, config.value);
+	}
+
+	modm::ResumableResult<bool>
+	setI2cAddress(uint8_t address)
+	{
+		return write(Register::I2C_ADDR, address);
+	}
+
+  /// Wait 1ms after setting the lower limit
+	modm::ResumableResult<bool>
+	setLowerLimit(Data data)
+	{
+		return write(Register::ZPOS, data.data);
+	}
+
+  /// Wait 1ms after setting the upper limit
+	modm::ResumableResult<bool>
+	setUpperLimit(Data data)
+	{
+		return write(Register::MPOS, data.data);
+	}
+
+	modm::ResumableResult<bool>
+	setMaxAngle(Data data)
+	{
+		return write(Register::MANG, data.data);
+	}
+
+  /// Permanently burn configurations
+	/// @warning As5600 can be burned only 3 times!
+	modm::ResumableResult<bool>
+	burn(Burn flags)
+	{
+		buffer[0] = static_cast<uint8_t>(flags);
+		return write(Register::BURN, buffer, 1);
+	}
+
+	modm::ResumableResult<Data>
+	getRawValue()
+	{
+		RF_BEGIN();
+
+		const uint16_t result = RF_CALL(read<uint16_t>(Register::ANGLE_RAW));
+
+		// Raw value requires masking
+		RF_END_RETURN(Data(result & Data::max));
+	}
+
+	modm::ResumableResult<Status>
+	getStatus()
+	{
+		return static_cast<Status>(read<uint8_t>(Register::STATUS));
+	}
+
+	modm::ResumableResult<uint16_t>
+	getMagnitude()
+	{
+		return read<uint16_t>(Register::MAGNITUDE);
+	}
+
+	/** Automated Gain Control
+	 *
+	 * The AS5600 uses Automatic Gain Control in a closed loop to compensate
+	 * for variations of the magnetic field strength due to changes of temperature,
+	 * airgap between IC and magnet, and magnet degradation.
+	 *
+	 * For the most robust performance, the gain value should be in the center of its range.
+	 * The airgap of the physical system can be adjusted to achieve this value.
+	 *
+	 * In 5V operation, range is 0-255
+	 * In 3.3V operation, range is reduced to 0-128
+	 */
+	modm::ResumableResult<uint8_t>
+	getAgcValue()
+	{
+		return read<uint8_t>(Register::AGC);
+	}
+
+	modm::ResumableResult<bool>
+	read()
+	{
+		RF_BEGIN();
+
+		data.data = RF_CALL(read<uint16_t>(Register::ANGLE));
+
+		RF_END_RETURN(this->wasTransactionSuccessful());
+	}
+
+	inline Data &
+	getData()
+	{
+		return data;
+	}
+
+private:
+	template<std::unsigned_integral T>
+	modm::ResumableResult<bool>
+	write(Register reg, T value)
+	{
+		RF_BEGIN();
+
+		buffer[0] = reg;
+
+		if constexpr (std::is_same_v<T, uint8_t>)
+		{
+			buffer[1] = value;
+		} else
+		{
+			buffer[1] = value >> 8;
+			buffer[2] = value;
+		}
+
+		this->transaction.configureWrite(buffer, 1 + sizeof(T));
+
+		return this->runTransaction();
```

Comment:
```suggestion
		return RF_END_RETURN_CALL(this->runTransaction());
```
Removing the macros sadly only works for a single function without any logic before or after. (Fibers are so much better ;-)

Reply:
😼 oops.. to much enthusiasm